src/ast/ast_validations.c

```diff
@@ -207,6 +207,69 @@ static AST_Validation _ValidateReferredFunctions(rax *referred_functions, char *
 	return res;
 }
 
+// Recursively collect function names and perform validations on functions with STAR arguments.
+static AST_Validation _VisitFunctions(const cypher_astnode_t *node, rax *func_names,
+									  char **reason) {
+	cypher_astnode_type_t type = cypher_astnode_type(node);
+	if(type == CYPHER_AST_APPLY_ALL_OPERATOR) {
+		// Working with a function call that has * as its argument.
+		const cypher_astnode_t *func = cypher_ast_apply_all_operator_get_func_name(node);
+		const char *func_name = cypher_ast_function_name_get_value(func);
+
+		// Verify that this is a COUNT call.
+		if(strcasecmp(func_name, "COUNT")) {
+			asprintf(reason, "COUNT is the only function which can accept * as an argument");
+			return AST_INVALID;
+		}
+
+		// Verify that DISTINCT is not specified.
+		if(cypher_ast_apply_all_operator_get_distinct(node)) {
+			// TODO consider opening a parser error, this construction is invalid in Neo's parser.
+			asprintf(reason, "Cannot specify both DISTINCT and * in COUNT(DISTINCT *)");
+			return AST_INVALID;
+		}
+
+		// Collect the function name.
+		raxInsert(func_names, (unsigned char *)func_name, strlen(func_name), NULL, NULL);
+
+		// As Apply All operators have no children, we can return here.
+		return AST_VALID;
+	}
+
+	if(type == CYPHER_AST_APPLY_OPERATOR) {
+		// Collect the function name.
+		const cypher_astnode_t *func = cypher_ast_apply_operator_get_func_name(node);
+		const char *func_name = cypher_ast_function_name_get_value(func);
+		raxInsert(func_names, (unsigned char *)func_name, strlen(func_name), NULL, NULL);
+	}
+
+	uint child_count = cypher_astnode_nchildren(node);
+	for(uint i = 0; i < child_count; i ++) {
+		const cypher_astnode_t *child = cypher_astnode_get_child(node, i);
+		AST_Validation res = _VisitFunctions(child, func_names, reason);
+		if(res != AST_VALID) return res;
+	}
+
+	return AST_VALID;
+}
+
+static AST_Validation _ValidateFunctionCalls(const cypher_astnode_t *node, char **reason,
```

Comment:
Shouldn't there be a call to `_ValidateFunctionCalls` from `_Validate_WITH_Clauses` and `_Validate_CREATE_Clauses`
`WITH none_existing_func() AS x, count(distinct *) AS z`
`CREATE ({x:none_existing_func(), z:count(distinct *)}) `

Reply:
:+1: Good call, we should have had similar calls all along.